File: examples/tests/vi/test_ffmpeg_demux.py
#!/usr/bin/env python3

import argparse
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import ff_pymedia as ff
from module_test_utils import RunMonitor, dump_output_channels


def parse_args():
    parser = argparse.ArgumentParser(description="Full ModuleFFmpegDemux usage test")
    parser.add_argument("input", help="media file, URL, or input device")
    parser.add_argument("--loop", type=int, default=1, help="loop count; -1 loops forever")
    parser.add_argument("--input-format", help="force FFmpeg input format")
    parser.add_argument(
        "--format-option", action="append", default=["probesize=200K"],
        metavar="KEY=VALUE", help="FFmpeg format option; may be repeated"
    )
    parser.add_argument("--timeout-us", type=int, default=5_000_000)
    parser.add_argument("--seek", type=int)
    parser.add_argument("--seek-flags", type=int, default=0)
    parser.add_argument("--buffers", type=int, default=20)
    parser.add_argument("--frames", type=int, default=0)
    parser.add_argument("--duration", type=float, default=0.0)
    parser.add_argument("--report-every", type=int, default=100)
    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--change-source")
    parser.add_argument("--external-consumer", action="store_true")
    parser.add_argument("--dump-pipe", action="store_true")
    return parser.parse_args()


def main():
    args = parse_args()
    monitor = RunMonitor(args.frames, args.duration, args.report_every, args.verbose)
    monitor.install_signal_handlers()

    demuxer = ff.ModuleFFmpegDemux(args.input, args.loop)
    if args.change_source:
        ret = demuxer.changeSource(args.change_source, args.loop)
        if ret < 0:
            print(f"changeSource failed: {ret}")
            return 1
    if args.input_format:
        ret = demuxer.setInputFormat(args.input_format)
        if ret < 0:
            print(f"setInputFormat failed: {ret}")
            return 1
    for item in args.format_option:
        key, separator, value = item.partition("=")
        if not separator or not key:
            print(f"Invalid format option: {item}")
            return 2
        ret = demuxer.setFormatOption(key, value, 0)
        if ret < 0:
            print(f"setFormatOption({key}) failed: {ret}")
            return 1
        print(f"Format option {key}={demuxer.getFormatOption(key, 0)}")
    demuxer.setTimeOut(args.timeout_us)
    demuxer.setBufferCount(args.buffers)
    demuxer.setMediaBufferProduceHooker(monitor.output_callback)
    demuxer.setMediaStatusChangeHooker(monitor.status_callback)

    ret = demuxer.init()
    if ret < 0:
        print(f"Failed to init demuxer: {ret}")
        return 1
    if args.seek is not None:
        ret = demuxer.setFileSeek(args.seek, args.seek_flags)
        if ret < 0:
            print(f"setFileSeek failed: {ret}")
            return 1

    dump_output_channels(demuxer)
    print(f"Video codec={demuxer.getVideoCodec()}, audio codec={demuxer.getAudioCodec()}")
    for media_type, label in ((ff.BUFFER_TYPE_VIDEO, "video"),
                              (ff.BUFFER_TYPE_AUDIO, "audio")):
        extra = demuxer.getExtraBuffer(media_type)
        print(f"{label} extra data: {extra.getActiveSize() if extra else 0} bytes")

    external = None
    external_frames = [0]
    if args.external_consumer:
        def external_callback(_name, _queue_size, _buffer):
            external_frames[0] += 1
        external = demuxer.addExternalConsumer("ffmpeg-demux-external", external_callback)
    if args.dump_pipe:
        demuxer.dumpPipe()

    monitor.reset()
    demuxer.start()
    monitor.wait()
    demuxer.stop()
    if args.dump_pipe:
        demuxer.dumpPipeSummary()
    monitor.print_summary("FFmpeg demuxer")
    if external is not None:
        print(f"External consumer received {external_frames[0]} buffers")
    return 1 if monitor.abnormal else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/tests/vi/test_ffmpeg_demux.cpp
#include "module/vi/module_ffmpegDemux.hpp"
#include "tests/module_test_utils.hpp"

#include <memory>
#include <sstream>

using namespace FFMedia;
using namespace FFMediaTest;

namespace
{
void usage(const char* program)
{
    ff_info("Usage: %s INPUT [options]\n"
            "  --loop 1                FFmpeg loop count (-1 forever)\n"
            "  --input-format NAME     Force AVInputFormat\n"
            "  --format-options K=V,K=V (default: probesize=200K)\n"
            "  --use-parameters        Configure through MediaParameter\n"
            "  --timeout-us 5000000 --seek TS --seek-flags N\n"
            "  --runtime-seek TS       Seek from the first produce callback\n"
            "  --buffers 20 --frames 0 --duration 0\n"
            "  --change-source INPUT   Exercise changeSource() before init\n"
            "  --external-consumer --verbose --report-every 100 --dump-pipe\n",
            program);
}

int setFormatOptions(ModuleFFmpegDemux& demuxer, const std::string& values)
{
    std::stringstream stream(values);
    std::string item;
    while (std::getline(stream, item, ',')) {
        const std::string::size_type equal = item.find('=');
        if (equal == std::string::npos || equal == 0) {
            ff_error("Invalid format option: %s\n", item.c_str());
            return -1;
        }
        const std::string key = item.substr(0, equal);
        const int ret = demuxer.setFormatOption(key, item.substr(equal + 1), 0);
        if (ret < 0)
            return ret;
        ff_info("Format option %s=%s\n", key.c_str(),
                demuxer.getFormatOption(key, 0).c_str());
    }
    return 0;
}
}  // namespace

int main(int argc, char** argv)
{
    CliOptions options(argc, argv);
    if (options.has("help") || options.positional().empty()) {
        usage(argv[0]);
        return options.has("help") ? 0 : 2;
    }

    const int loop = options.getLong("loop", 1);
    const bool use_parameters = options.getBool("use-parameters");
    auto demuxer = std::make_shared<ModuleFFmpegDemux>(
        use_parameters ? std::string() : options.positional()[0],
        use_parameters ? 1 : loop);
    if (use_parameters) {
        int ret = demuxer->setParameter("source/uri", options.positional()[0]);
        if (ret == 0)
            ret = demuxer->setParameter("source/loop", loop);
        if (ret < 0) {
            ff_error("Setting FFmpeg source parameters failed: %d\n", ret);
            return ret;
        }
    }
    if (options.has("change-source")) {
        const int ret = demuxer->changeSource(options.get("change-source"), loop);
        if (ret < 0) {
            ff_error("changeSource failed: %d\n", ret);
            return ret;
        }
    }
    if (options.has("input-format")) {
        const int ret = use_parameters
                            ? demuxer->setParameter(
                                "input-format", options.get("input-format"))
                            : demuxer->setInputFormat(
                                options.get("input-format"));
        if (ret < 0) {
            ff_error("setInputFormat failed: %d\n", ret);
            return ret;
        }
    }
    const std::string format_options = options.get("format-options", "probesize=200K");
    int ret = use_parameters
                  ? demuxer->setParameter("ffmpeg/options", format_options)
                  : setFormatOptions(*demuxer, format_options);
    if (ret < 0)
        return ret;
    if (use_parameters) {
        ret = demuxer->setParameter(
            "timeout-usec", options.getLong("timeout-us", 5000000));
        if (ret == 0)
            ret = demuxer->setParameter(
                "buffer-count", options.getLong("buffers", 20));
        if (ret < 0)
            return ret;
    } else {
        demuxer->setTimeOut(options.getLong("timeout-us", 5000000));
        demuxer->setBufferCount(options.getLong("buffers", 20));
    }

    RunMonitor monitor(options.getLong("frames", 0),
                       options.getDouble("duration", 0.0),
                       options.getLong("report-every", 100),
                       options.getBool("verbose"));
    installSignalHandlers(monitor);
    std::atomic<bool> runtime_seek_attempted(false);
    std::atomic<int> runtime_seek_result(0);
    ModuleFFmpegDemux* demuxer_ptr = demuxer.get();
    demuxer->setMediaBufferProduceHooker(
        [&monitor, &options, use_parameters, demuxer_ptr,
         &runtime_seek_attempted, &runtime_seek_result](
            const std::string& name, int queue_size,
            std::shared_ptr<MediaBuffer> buffer) {
            if (options.has("runtime-seek")
                && !runtime_seek_attempted.exchange(true)) {
                int ret = 0;
                if (use_parameters) {
                    ret = demuxer_ptr->setParameter(
                        "seek-flags", options.getLong("seek-flags", 0));
                    if (ret == 0) {
                        ret = demuxer_ptr->setParameter(
                            "seek", options.getLong("runtime-seek", 0));
                    }
                } else {
                    ret = demuxer_ptr->setFileSeek(
                        options.getLong("runtime-seek", 0),
                        options.getLong("seek-flags", 0));
                }
                runtime_seek_result.store(ret);
                if (ret < 0)
                    monitor.requestStop();
            }
            monitor.onBuffer(name, queue_size, buffer);
        });
    demuxer->setMediaStatusChangeHooker(
        [&monitor](const std::string& name, MediaStatus status) {
            monitor.onStatus(name, status);
        });

    ret = demuxer->init();
    if (ret < 0) {
        ff_error("Failed to init FFmpeg demuxer: %d\n", ret);
        return ret;
    }
    if (use_parameters) {
        int64_t seek_max = 0;
        bool seekable = false;
        ret = demuxer->getParameter("seek-max", seek_max);
        if (ret == 0)
            ret = demuxer->getParameter("seekable", seekable);
        if (ret < 0) {
            ff_error("Querying FFmpeg seek parameters failed: %d\n", ret);
            return ret;
        }
        ff_info("FFmpeg seek max=%ld us, seekable=%d\n",
                seek_max, seekable);
    }
    if (options.has("seek")) {
        if (use_parameters) {
            ret = demuxer->setParameter(
                "seek-flags", options.getLong("seek-flags", 0));
            if (ret == 0)
                ret = demuxer->setParameter(
                    "seek", options.getLong("seek", 0));
        } else {
            ret = demuxer->setFileSeek(options.getLong("seek", 0),
                                       options.getLong("seek-flags", 0));
        }
        if (ret < 0) {
            ff_error("FFmpeg seek failed: %d\n", ret);
            return ret;
        }
    }

    dumpOutputMediaChannels(*demuxer);
    ff_info("Video codec=%d, audio codec=%d\n",
            static_cast<int>(demuxer->getVideoCodec()),
            static_cast<int>(demuxer->getAudioCodec()));
    dumpExtraBuffer("video", demuxer->getExtraBuffer(BUFFER_TYPE_VIDEO));
    dumpExtraBuffer("audio", demuxer->getExtraBuffer(BUFFER_TYPE_AUDIO));

    std::shared_ptr<ModuleMedia> external_consumer;
    std::atomic<uint64_t> external_frames(0);
    if (options.has("external-consumer")) {
        external_consumer = demuxer->addExternalConsumer(
            "ffmpeg-demux-external", [&external_frames](const std::string&, int,
                                                        std::shared_ptr<MediaBuffer>) {
                ++external_frames;
            });
    }
    if (options.has("dump-pipe"))
        demuxer->dumpPipe();

    monitor.reset();
    demuxer->start();
    monitor.wait();
    demuxer->stop();
    if (options.has("dump-pipe"))
        demuxer->dumpPipeSummary();
    monitor.printSummary("FFmpeg demuxer");
    if (external_consumer)
        ff_info("External consumer received %lu buffers\n", external_frames.load());
    if (options.has("runtime-seek")
        && (!runtime_seek_attempted.load()
            || runtime_seek_result.load() < 0)) {
        return 1;
    }
    return monitor.abnormal() ? 1 : 0;
}
